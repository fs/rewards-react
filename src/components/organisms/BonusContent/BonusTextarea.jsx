import React from 'react';
import ReactTextareaAutocomplete from '@webscopeio/react-textarea-autocomplete';
import '../../../styles/autocomplete.css';
import BonusPossibilitiesService, {
  POINTS, TAGS, USERS,
} from '../../../services/BonusPossibilitiesService';

const BonusTextarea = (props) => {
  const { onChange } = props;
  const pointItem = ({ entity: { id, value } }) => <div data-testid="test-point-item" key={id}>{`+â™¥${value}`}</div>;
  const userItem = ({ entity: { id, username } }) => <div data-testid="test-user-item" key={id}>{`@${username}`}</div>;
  const tagItem = ({ entity: { id, label } }) => <div data-testid="test-tag-item" key={id}>{`#${label}`}</div>;

  const points = BonusPossibilitiesService.getPossibilities(POINTS);
  const users = BonusPossibilitiesService.getPossibilities(USERS);
  const tags = BonusPossibilitiesService.getPossibilities(TAGS);

  return (
    <div data-testid="test-textarea-wrapper">
      <ReactTextareaAutocomplete
        className="autocomplete-textarea"
        onChange={onChange}
        placeholder="+100 @person add description for #create_awesomness"
        data-testid="test-textarea"
        loadingComponent={() => <span>Loading</span>}
        name="bonus-text"
        trigger={{
          '+': {
            dataProvider: token => points.filter(point => point.id.includes(token)),
            component: pointItem,
            output: (item, trigger) => `${trigger}${(item.value).toString()}`,
          },
          '@': {
            dataProvider: token => users.filter(user => user.username.includes(token)),
            component: userItem,
            output: (item, trigger) => `${trigger}${item.username}`,
          },
          '#': {
            dataProvider: token => tags.filter(tag => tag.label.includes(token)),
            component: tagItem,
            output: (item, trigger) => `${trigger}${item.label}`,
          },
        }}
        minChar="0"
      />
    </div>
  );
};

export default BonusTextarea;
