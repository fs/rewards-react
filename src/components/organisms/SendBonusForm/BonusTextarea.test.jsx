import React from 'react';
import { shallow } from 'enzyme';
import ReactTextareaAutocomplete from '@webscopeio/react-textarea-autocomplete';

import BonusTextarea, { pointItem, userItem, tagItem } from './BonusTextarea';

describe('BonusTextarea', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  test('should show tags if user type #', () => {
    // Arrange
    const expectedPointChar = '+';
    const expectedUserChar = '@';
    const expectedTagChar = '#';

    // Act
    const wrapper = shallow(<BonusTextarea onChange={() => {}} />);

    const actualPointTrigger = Object.getOwnPropertyDescriptor(
      wrapper.find(ReactTextareaAutocomplete).props().trigger,
      expectedPointChar,
    );
    const actualUserTrigger = Object.getOwnPropertyDescriptor(
      wrapper.find(ReactTextareaAutocomplete).props().trigger,
      expectedUserChar,
    );
    const actualTagTrigger = Object.getOwnPropertyDescriptor(
      wrapper.find(ReactTextareaAutocomplete).props().trigger,
      expectedTagChar,
    );

    // Assert
    expect(wrapper.find(ReactTextareaAutocomplete)).toHaveLength(1);
    expect(actualPointTrigger.value.component).toEqual(pointItem);
    expect(actualUserTrigger.value.component).toEqual(userItem);
    expect(actualTagTrigger.value.component).toEqual(tagItem);
  });
});
