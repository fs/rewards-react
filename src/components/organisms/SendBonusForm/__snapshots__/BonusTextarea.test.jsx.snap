// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BonusTextarea Autocomplete Items pointItem 1`] = `
<div
  data-testid="test-point-item"
>
  +â™¥25
</div>
`;

exports[`BonusTextarea Autocomplete Items tagItem 1`] = `
<div
  data-testid="test-tag-item"
>
  #win-win-win
</div>
`;

exports[`BonusTextarea Autocomplete Items userItem 1`] = `
<div
  data-testid="test-user-item"
>
  @albert.fazullin
</div>
`;
